grpc/server.go

```diff
@@ -0,0 +1,251 @@
+ * identify the disruption to apply to a query. The mapping represents user preference the proportion of queries
+ * affected by each alteration. See docs/grpc_disruption/interceptor_algorithm.md for examples.
+ */
+
+// DisruptionListener is a gRPC Service that can disrupt endpoints of a gRPC server.
+type DisruptionListener struct {
+	pb.UnimplementedDisruptionListenerServer
+	Configuration DisruptionConfiguration
+	Logger        *zap.SugaredLogger
+}
+
+// DisruptionConfiguration configures the DisruptionListener to chaos test endpoints of a gRPC server.
+type DisruptionConfiguration map[TargetEndpoint]EndpointConfiguration
+
+// EndpointConfiguration configures endpoints that the DisruptionListener chaos tests on a gRPC server.
+// The AlterationHash maps integers from 0 to 100 to alteration configurations.
+type EndpointConfiguration struct {
+	TargetEndpoint TargetEndpoint
+	Alterations    map[AlterationConfiguration]PercentAffected
+	AlterationHash []AlterationConfiguration
+}
+
+// AlterationConfiguration contains either an ErrorToReturn or an OverrideToReturn for a given
+// gRPC query to the disrupted service.
+type AlterationConfiguration struct {
+	ErrorToReturn    string
+	OverrideToReturn string
+}
+
+// PercentAffected is an integer represented the number of queries out of 100 queries which should
+// be affected by a certain endpoint alteration.
+type PercentAffected int
+
+// TargetEndpoint is a string of the format /package.service/method.
+type TargetEndpoint string
+
+var mutex sync.Mutex
+
+// NewDisruptionListener creates a new DisruptionListener Service with the logger instantiated and DisruptionConfiguration set to be empty
+func NewDisruptionListener() (*DisruptionListener, error) {
+	d := DisruptionListener{}
+
+	var err error
+	d.Logger, err = log.NewZapLogger()
+
+	d.Configuration = DisruptionConfiguration{}
+
+	return &d, err
+}
+
+// SendDisruption receives a disruption specfication and configures the interceptor to spoof responses to specified endpoints.
+func (d *DisruptionListener) SendDisruption(ctx context.Context, ds *pb.DisruptionSpec) (*emptypb.Empty, error) {
+	if ds == nil {
+		d.Logger.Error("Cannot execute SendDisruption when DisruptionSpec is nil")
+		return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption when DisruptionSpec is nil")
+	}
+
+	config := DisruptionConfiguration{}
+
+	for _, endpointSpec := range ds.Endpoints {
+		if endpointSpec.TargetEndpoint == "" {
+			d.Logger.Error("DisruptionSpec does not specify TargetEndpoint for at least one endpointAlteration")
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption without specifying TargetEndpoint for all endpointAlterations")
+		}
+
+		alterationToPercentAffected, err := GetAlterationToPercentAffected(endpointSpec.Alterations)
+		if err != nil {
+			return nil, err
+		}
+
+		percentToAlteration := GetPercentToAlteration(alterationToPercentAffected, d.Logger)
```

Comment:
If the output of `GetAlterationToPercentAffected` is sent directly into `GetPercentToAlteration` before being added to the config, why have two separate functions?